### PR TITLE
zigHook: support zig 0.11

### PR DESCRIPTION
**Pull request by figsoda**
## Description of changes

also added a package for testing

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/zig/hook.nix [hunk 1]
```diff
@@ -8,6 +8,14 @@ makeSetupHook {
 
   propagatedBuildInputs = [ zig ];
 
+  substitutions = {
+    zig_default_flags =
```
Comment (AndersonTorres) [inline pkgs/development/compilers/zig/hook.nix:12]: Is this passed as one string, not split in whitespaces?
That's strange...

File: pkgs/development/compilers/zig/setup-hook.sh [hunk 2]
```diff
@@ -20,8 +20,6 @@
 # https://github.com/NixOS/nixpkgs/pull/241741#issuecomment-1624227485
 # https://github.com/ziglang/zig/issues/14281#issuecomment-1624220653
 
-readonly zigDefaultFlagsArray=("-Drelease-safe=true" "-Dcpu=baseline")
```
Comment (AndersonTorres) [inline pkgs/development/compilers/zig/setup-hook.sh:23]: Instead of spreading the macro substitution around all the text, it is better to concentrate it here:

```bash
readonly zigDefaultFlagsArray=(@zig_default_flags@)
```

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -3006,6 +3006,12 @@ with pkgs;
 
   pikchr = callPackage ../tools/graphics/pikchr { };
 
+  poop = callPackage ../tools/misc/poop {
+    zigHook = zigHook.override {
+      zig = zig_0_11;
```
Comment (AndersonTorres) [inline pkgs/top-level/all-packages.nix:3011]:
```suggestion
      zig = buildPackages.zig_0_11;
```

Cross-compilation magic.

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/compilers/zig/hook.nix (hunk 1 above)
Comment (figsoda) [reply pkgs/development/compilers/zig/hook.nix:12]: this is just plain text substitution with `substituteAll`

Comment (AndersonTorres) [reply pkgs/development/compilers/zig/hook.nix:12]: This is because when I created such arrays the whitespaces were `\`-quoted, generating some nasty bugs.

**Review by AndersonTorres — APPROVED** [review]
LGTM.

**Comment by Atemu** [comment]
💩

**Comment by sagehane** [comment]
There's a flaw in this approach to support Zig 0.11. If the `.preferred_optimize_mode` field is defined to something, the build system won't accept `-Drelease-<safe>=true` and instead only accept `-Doptimize=true`.

Using the `zigHook` for such projects would fail as Zig rejects unknown flags. Thankfully, I don't think any Zig package in Nixpkgs defines this field yet.

Also, not all projects have their preferred optimize mode as Safe. As such, I think it would be best to:
1. Set the default optimisation flag is set to `-Doptimize=true` and make it easily override-able.
2. Have the maintainer override the default optimisation flag on a per-package basis when the upstream doesn't provide one.

**Comment by figsoda** [comment]
Currently, `zigHook` makes a lot of assumptions about `build.zig` that are hard to avoid until we see something like [this](https://github.com/ziglang/zig/issues/14281#issuecomment-1624220653)

> 1. Set the default optimisation flag is set to `-Doptimize=true`

The default `build.zig` generated by `zig init-exe` doesn't accept `-Doptimize=true`, so I think `-Doptimize=ReleaseSafe` is a better default

> and make it easily override-able.

This is hard because `zigHook` is used for both zig >= 0.11 and zig <= 0.10, which has different standard optimization flags (e.g. `-Doptimize=ReleaseSafe` vs `-Drelease-safe=true`, it would take some amount of guess work and string manipulation for the same setting to work across zig versions, not to mention that we may not want this option to exist in the long term given the possibility of something like `--release`

> Have the maintainer override the default optimisation flag on a per-package basis when the upstream doesn't provide one.

This sounds like more work than using zig directly without the hook for packages that don't use the default options.

**Comment by AndersonTorres** [comment]
As I said before, all those things depend on some upstream work. Indeed, doing a bit of reverse-futurology, I believe that if it wasn't for such weird distros as Guix and Nix, such issues of [r13y](https://r13y.com/) would go unnoticed for decades...

**Comment by sagehane** [comment]
tl;dr - At the very least, there needs to be an easy to override the optimisation mode flag in order to prevent future compile errors with projects defining their preferred optimisation mode. As for the default, I don't think it's worth defaulting to safe mode but I won't complain too much.

> > 1. Set the default optimisation flag is set to `-Doptimize=true`
> 
> The default `build.zig` generated by `zig init-exe` doesn't accept `-Doptimize=true`, so I think `-Doptimize=ReleaseSafe` is a better default

The issue with defaulting to safe mode is that there's no guarantee that that's what the upstream expects, so the packager needs to check the upstream project for their packaging guidelines anyways. It feels like a minor footgun that could be avoided.

The appeal of defaulting to `-Doptimize=true` is that *when it works, it's likely to be correct.* My idea of good defaults are defaults that are silent when correct and complain when misapplied, what do you think?

> > and make it easily override-able.
> 
> This is hard because `zigHook` is used for both zig >= 0.11 and zig <= 0.10, which has different standard optimization flags (e.g. `-Doptimize=ReleaseSafe` vs `-Drelease-safe=true`, it would take some amount of guess work and string manipulation for the same setting to work across zig versions, not to mention that we may not want this option to exist in the long term given the possibility of something like `--release`

Well, this is pretty much required either way because lots of builds will fail if the project does define `.preferred_optimize_mode`. When packaging for such projects, people will either have to override the flag or patch the build system.

If overriding the default flag becomes the norm for many projects, it would also make it easier to support older versions without using string magic.

> > Have the maintainer override the default optimisation flag on a per-package basis when the upstream doesn't provide one.
> 
> This sounds like more work than using zig directly without the hook for packages that don't use the default options.

I just disagree that defaulting to safe mode is a default worth depending on.
